### PR TITLE
Expose individual partitions

### PR DESCRIPTION
**Pull request by dim**
Addressing #162 and #147

### CODE REVIEW

File: config.go [hunk 1]
```diff
@@ -9,15 +9,32 @@ import (
...
 	sarama.Config
 
 	// Group is the namespace for group management properties
 	Group struct {
+
 		// The strategy to use for the allocation of partitions to consumers (defaults to StrategyRange)
 		PartitionStrategy Strategy
-		Offsets           struct {
+
+		// By default, messages and errors from the subsribed topics and partitions are all multiplexed and
```
Comment (jgeiger) [inline config.go]: subscribed

**Comment by jgeiger** [comment]
It's funny this appears right as I need it. The only thing that I had to do with #162 was to deal with making sure on re-balance that the goroutine I was using to read the partition was properly destroyed when the app was no longer responsible for that partition. Not hard, but it might be useful to send a message/signal or ? when the partition is no longer handled by the app.

Adding seemed to work pretty well, I just didn't want to leave orphans. It also may have been the way I chose to process the messages (select vs range) since I wanted to close the goroutines on context.Done()

**Comment by dim** [comment]
@jgeiger I have updated the example consumer. The `Messages()` and the `Errors()` channels will be closed when the `PartitionConsumer` is closed which means that you can use them as your signal to exit the goroutine

**Comment by jgeiger** [comment]
Thanks. You get so used to seeing things one way (val, ok) that it just forget. I think the examples are good now for the use case.

File: README.md [hunk 2]
```diff
@@ -68,6 +71,78 @@ func main() {
...
+					fmt.Fprintf(os.Stdout, "%s/%d/%d\t%s\t%s\n", msg.Topic, msg.Partition, msg.Offset, msg.Key, msg.Value)
+					consumer.MarkOffset(msg, "")	// mark message as processed
+				case err, more := <-consumer.Errors():
+					if !more {	// exit when the partition is closed
+						return
+					}
+					log.Printf("Error: %s\n", err.Error())
+				}
+			}(part)
+		case ntf, more := <-consumer.Notifications():
```
Comment (crast) [inline README.md]: I know it's just an example, but what's the advantage of putting everything in a giant select?

I think it'd be equally useful to `for range` loop Notifications in another goroutine... after all it should close once the consumer does.. makes the example more concise and less indented.

File: examples_test.go [hunk 3]
```diff
@@ -52,3 +52,61 @@ func ExampleConsumer() {
...
+			// start a separate goroutine to consume the partition;
+			go func(pc cluster.PartitionConsumer) {
+				select {
+				case msg, more := <-pc.Messages():
+					if !more { // exit when the partition is closed
+						return
+					}
+					fmt.Fprintf(os.Stdout, "%s/%d/%d\t%s\t%s\n", msg.Topic, msg.Partition, msg.Offset, msg.Key, msg.Value)
+					consumer.MarkOffset(msg, "") // mark message as processed
+				case err, more := <-consumer.Errors():
```
Comment (crast) [inline examples_test.go]: Here you have N instances of selecting over consumer.Errors which won't break, but is not really necessary.. and is not semantically correct as these errors are not partition-specific.

I'd move this outside here to alongside `consumer.Notifications`

This also has the benefit of letting the Messages consumer be a simple `for range` loop

File: README.md (hunk 2 above)
Comment (crast) [reply README.md]: Modified example: https://gist.github.com/crast/2f179d88d169cea7e5deadb7f11d593f

Comment (dim) [reply README.md]: well, preferences, but I agree, the giant select is not pretty

File: examples_test.go (hunk 3 above)
Comment (dim) [reply examples_test.go]: so this is one I think I was wondering about, should I expose individual Errors channels or should I continue to multiplex them? it doesn't seem like this is performance critical and it's certainly more convenient to consume them from a single source.

**Comment by dim** [comment]
@crast can you please have one more look?

**Comment by crast** [comment]
@dim can you release a new version tag, for those of us using `dep` ?

**Comment by dim** [comment]
@crast absolutely, it's on my TODO list, just trying to sort out #174 first. In the meantime, pls use `revision = "xxx"` in your Gopkg.toml

**Pull request merged**